Source/Core/Core/IOS/USB/Bluetooth/LibUSBBluetoothAdapter.cpp

```diff
@@ -0,0 +1,574 @@
+  return buffer;
+}
+
+auto LibUSBBluetoothAdapter::ReceiveACLData() -> BufferType
+{
+  BufferType buffer;
+  m_acl_data_queue.Pop(buffer);
+  return buffer;
+}
+
+bool LibUSBBluetoothAdapter::IsControllerReadyForCommand() const
+{
+  return m_num_hci_command_packets > m_unacknowledged_commands.size();
+}
+
+template <typename EventType>
+void LibUSBBluetoothAdapter::AcknowledgeCommand(std::span<const u8> buffer)
+{
+  if (buffer.size() < sizeof(hci_event_hdr_t) + sizeof(EventType)) [[unlikely]]
+  {
+    WARN_LOG_FMT(IOS_WIIMOTE, "Undersized HCI event");
+    return;
+  }
+
+  EventType ev;
+  std::memcpy(&ev, buffer.data() + sizeof(hci_event_hdr_t), sizeof(ev));
+
+  const auto it =
+      std::ranges::find(m_unacknowledged_commands, ev.opcode, &OutstandingCommand::opcode);
+  if (it != m_unacknowledged_commands.end())
+  {
+    DEBUG_LOG_FMT(IOS_WIIMOTE, "HCI command acknowledged: 0x{:04x}", ev.opcode);
+    m_unacknowledged_commands.erase(it);
+  }
+  else if (ev.opcode != 0x0000)
+  {
+    WARN_LOG_FMT(IOS_WIIMOTE, "Unexpected opcode acknowledgement: 0x{:04x}", ev.opcode);
+  }
+
+  m_num_hci_command_packets = ev.num_cmd_pkts;
+}
+
+libusb_transfer* LibUSBBluetoothAdapter::AllocateTransfer(std::size_t buffer_size)
+{
+  auto& item =
+      *m_transfer_buffers.emplace(libusb_alloc_transfer(0), Common::UniqueBuffer<u8>{buffer_size})
+           .first;
+
+  auto* const transfer = item.first;
+  transfer->buffer = item.second.get();
+
+  return transfer;
+}
+
+void LibUSBBluetoothAdapter::FreeTransfer(libusb_transfer* transfer)
+{
+  libusb_free_transfer(transfer);
+  m_transfer_buffers.erase(transfer);
+}
+
+void LibUSBBluetoothAdapter::CleanCompletedTransfers()
+{
+  libusb_transfer* transfer{};
+  while (m_completed_transfers.Pop(transfer))
+    FreeTransfer(transfer);
+}
+
+void LibUSBBluetoothAdapter::PushHCICommand(TimedTransfer transfer)
+{
+  hci_cmd_hdr_t cmd;
+  std::memcpy(&cmd, libusb_control_transfer_get_data(transfer.transfer), sizeof(cmd));
+
+  // Push the time forward so our timeouts work properly if command was queued and delayed.
+  const auto submit_time = std::max(transfer.target_time, Clock::now());
+
+  m_unacknowledged_commands.emplace_back(OutstandingCommand{cmd.opcode, submit_time});
+
+  --m_num_hci_command_packets;
+
+  m_output_worker.EmplaceItem(transfer);
+}
+
+void LibUSBBluetoothAdapter::SubmitTimedTransfer(TimedTransfer transfer)
+{
+  if (Config::Get(Config::MAIN_PRECISION_FRAME_TIMING))
+    m_precision_timer.SleepUntil(transfer.target_time);
+  else
+    std::this_thread::sleep_until(transfer.target_time);
+
+  SubmitTransfer(transfer.transfer);
+}
+
+void LibUSBBluetoothAdapter::SubmitTransfer(libusb_transfer* transfer)
+{
+  const int ret = libusb_submit_transfer(transfer);
+
+  if (ret == LIBUSB_SUCCESS)
+    return;
+
+  ERROR_LOG_FMT(IOS_WIIMOTE, "libusb_submit_transfer: {}", LibusbUtils::ErrorWrap(ret));
+
+  std::lock_guard lk{m_transfers_mutex};
+  m_completed_transfers.Emplace(transfer);
```

Comment:
Nitpicking but thoughts on calling this `m_transfers_to_free` ?  Alternatively, maybe add a comment like:

```
// Submission failed, cleanup this transfer on the next Update
```

Also, why do we need to do this for failed transfers but not successful ones?  I would think we don't do it for any of them because the transfer is in `m_transfer_buffers` ?

Reply:
Yeah, that rename makes sense. Successful transfers push to `m_completed_transfers` in the libusb callback after actually using the buffer. Failed transfers don't invoke the callback so they need to be marked for release here. I'll add a comment.

Reply:
Ahhh, sorry, I didn't follow it all the way through.  That makes sense.  Thanks for the details.